### PR TITLE
feat: add custom sources

### PR DESCRIPTION
**Pull request by talwat**
I'm opening this PR since I want to get some input about the documentation of this feature and so on before it's merged. So let me know :)

This adds custom track lists, which can be used with `--tracks`. The code still needs some review though, so if you're willing to review it, it would be extremely appreciated.

### CODE REVIEW

**Comment by danielwerg** [comment]
- Track progress is broken and total lengths is 00:00 [1]
- https://lofigirl.com/wp-content/uploads/2023/06/01-Sample-Track.mp3 returns 404, makes played stuck on `loading` state if this track is played first

archive.org is still down ;(

[1]
image

**Comment by danielwerg** [comment]
Tried to play tracks from my local folder:

```
python -m http.server 8088
```

```
lowfi --tracks ~/list.txt -d
```

~/list.txt:
```
0.0.0.0:8088/
x.mp3
```

No error message just infinite `loading` state

But if I open http://0.0.0.0:8088/x.mp3 in browser it plays just fine.

**Comment by talwat** [comment]
> Would it also be possible for lowfi to just search for *.mp3 files inside base url instead of making user create a list of tracks?

No, because pretty much 0 websites willingly hand over a list of all their mp3 files. It's possible, but faaar out of scope for lowfi.

**Comment by talwat** [comment]
> Would it also be possible for lowfi to just search for *.mp3 files inside base url instead of making user create a list of tracks?

No, because pretty much 0 websites willingly hand over a list of all their mp3 files. It's possible, but faaar out of scope for lowfi.

**Comment by talwat** [comment]
> - Track progress is broken and total lengths is 00:00 [1]

Ah... That's something I thought might happen. I'll try and fix it, but some mp3 files don't support getting the total duration for some reason.

> - https://lofigirl.com/wp-content/uploads/2023/06/01-Sample-Track.mp3 returns 404, makes played stuck on `loading` state if this track is played first

Yeah, this is again just me being a dumbass, I'll fix it soon :)

> archive.org is still down ;(

This I can't do anything about, we'll have to wait for that :(

**Comment by talwat** [comment]
Update on the "infinite loading state",

You could try:

```txt
http://0.0.0.0:8088/
x.mp3
```

Since lowfi only supports `http` and `https`, it isn't really designed for local files unless you're willing to make an http fileserver with them.

**Comment by danielwerg** [comment]
Okay this worked :O

```diff
- 0.0.0.0:8088/
+ http://0.0.0.0:8088/
x.mp3
```

---

Had a problem with non-latin letters or special symbols

File name: `TEST - HELLOФФФФ.mp3`
image
File name: `ヅ`
image
File name: `ＭＹ ＫＩＮＤ.mp3`
image

---

I also noticed lowfi formats file name before displaying it as track name. Often they don't looks so pretty. I can decide on one convention inside my music library for file names but I don't want to make it specifically for lowfi. It will be "ARTIST - TRACK". Would be great if user somehow could provide parsing method for filename to lowfi, I would just split by ` - ` and get [1]. Also I wouldn't change letter casing and keep as is while it's good for general use but it's not for my library.

Default lowfi parser:
`"hELLo - worlD" -> "Hello World"`
My imagined custom parsed:
`"hELLo - worlD" -> "worlD"`

**Comment by talwat** [comment]
I've gone ahead and fixed the special character bug with https://github.com/talwat/lowfi/commit/59cef9df6f296b9fbae013c3c5c351e11ca485d4, but I think the formatting is just too specific to implement, since it's a relatively minor thing that would require more effort than it's worth.

**Comment by danielwerg** [comment]
Hmmm, https://github.com/talwat/lowfi/commit/59cef9df6f296b9fbae013c3c5c351e11ca485d4 seems to done nothing for me

---

As for formatting, you are right. But can we then set track name inside tracks.txt file optionally? Make lowfi read between specific symbols or similar.

In my use case I would have to start file server in my music folder and then I will have a script that will generate `tracks.txt` file from files inside that music dir I mentioned above to feed to lowfi. I can modify my script to parse out desired track name and put it to `tracks.txt` for lowfi to read if this is implemented.

**Comment by talwat** [comment]
Ah yep you're right, https://github.com/talwat/lowfi/commit/21c0e54b66f4160d501855bb9ef541aea87a74e5 should do the trick.

About custom naming, I again am hesitant about it, but I will look into it to see if it needs major restructuring or if it's fairly quick to implement. I might do something like `<track url>!<track name>` or something else similar to that.

**Comment by danielwerg** [comment]
Hmmm, I don't think https://github.com/talwat/lowfi/commit/21c0e54b66f4160d501855bb9ef541aea87a74e5 changed anything with file names I gave above. So second section in https://github.com/talwat/lowfi/pull/21#issuecomment-2408616068 is still relevant.

This made me think how lowfi would look without border :D

**Comment by talwat** [comment]
> Hmmm, I don't think https://github.com/talwat/lowfi/commit/21c0e54b66f4160d501855bb9ef541aea87a74e5 changed anything with file names I gave above. So second section in https://github.com/talwat/lowfi/pull/21#issuecomment-2408616068 is still relevant.

It did, since it's using `.chars().size_hint()` on the actual track name, since before it was using it on the status (`paused`, `playing`, etc...)

**Comment by danielwerg** [comment]
Now it's other way around:
image

Same for `ヅ`, but `TEST - HELLOФФФФ.mp3` is fixed.

**Comment by talwat** [comment]
> Same for `ヅ`, but `TEST - HELLOФФФФ.mp3` is fixed.

Oh boy...

**Comment by talwat** [comment]
I think I've figured out the issue, roughly.

Some graphemes (unicode characters) are _wider_ than others, meaning I need to not only count for graphemes but also figure out how long each grapheme is.

So this will take a while, if you have any suggestions let me know.

**Comment by talwat** [comment]
Well, I've switched crates, so it might work now.

**Comment by danielwerg** [comment]
It does! I tested on all three I mentioned on https://github.com/talwat/lowfi/pull/21#issuecomment-2408616068.

**Comment by talwat** [comment]
Fantastic! I think I'll do one last code review later today and then merge this PR.

As for the custom track names, I think that'll be it's own PR, since it shouldn't be a breaking change to existing tracklists.

File: README.md [hunk 1]
```diff
@@ -96,3 +96,42 @@ An example of scrape is as follows,
...
+2023/06/Foudroie-Finding-The-Edge-V2.mp3
+2023/04/2-In-Front-Of-Me.mp3
+https://file-examples.com/storage/fea570b16e6703ef79e65b4/2017/11/file_example_MP3_5MG.mp3
+```
+
+lowfi would download these three URLs:
+
+- `https://lofigirl.com/wp-content/uploads/2023/06/Foudroie-Finding-The-Edge-V2.mp3`
+- `https://file-examples.com/storage/fea570b16e6703ef79e65b4/2017/11/file_example_MP3_5MG.mp3`
+- `https://lofigirl.com/wp-content/uploads/2023/04/2-In-Front-Of-Me.mp3`
```
Comment (danielwerg) [inline README.md:146]: Shouldn't line `145` and `146` be swap here? Or am I missing something?

**Review by danielwerg — APPROVED** [review]

File: README.md (hunk 1 above)
Comment (talwat) [reply README.md:146]: The order is irrelevant, since it's randomized either way.

**Pull request merged**